### ISSUE TITLE
gdb "cannot subscript non-array type" to index a Vec

### ISSUE THREAD

**Issue by Hi-Angel** [rust-lang/rust#66482]
Attempt to print content of a `Vec` at index results in error. Since it works fine for C++, I think the problem is in `rustc` not providing the necessary debug information for this to work.

`Vec`s are a widely used type, and not being able to print their content at index hurts debugging process noticeably.

## Steps to reproduce *(in terms of terminal commands)*

```
$ cat -n test2.rs
     1  fn main() {
     2      let x: Vec<usize> = vec![1,2,3];
     3      println!("{:?}", x);
     4  }
$ rustc test2.rs -o a -g
$ gdb ./a
Reading symbols from ./a...
warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
of file /tmp/a.
Use `info auto-load python-scripts [REGEXP]' to list them.
gdb λ br 3
Breakpoint 1 at 0x5c05: file test2.rs, line 3.
gdb λ r
Starting program: /tmp/a
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, test2::main () at test2.rs:3
3           println!("{:?}", x);
gdb λ p x[0]
```

### Expected

A print:

    1

### Actual

It prints:

    Cannot subscript non-array type

## Additional information

rustc version: `rustc 1.41.0-nightly (1bd30ce2a 2019-11-15)`
gdb version: `8.3.1`

**Comment by Hi-Angel** [comment]
Probably related: https://github.com/rust-lang/rust/issues/33014

**Comment by tromey** [comment]
See also https://sourceware.org/bugzilla/show_bug.cgi?id=21178

**Comment by Hi-Angel** [comment]
> See also https://sourceware.org/bugzilla/show_bug.cgi?id=21178

Thanks! So, to be clear: is it a bug in rust compiler or in gdb? I feel like one of the reports *(either this one, or the referenced one)* should be closed.

**Comment by JerrieLau** [comment]
lldb is fine.

**Comment by Hi-Angel** [comment]
> lldb is fine.

Thanks, nice to know! I took a look at it, one disappointing thing in comparison to gdb is that it doesn't have a way to enter python commands. But otherwise, judging by [gdb to lldb cmds map](https://lldb.llvm.org/use/map.html) it looks to be a nice alternative, worth to bear in mind.

**Comment by tromey** [comment]
> lldb is fine.

I'm very surprised to hear this.  How does it work?

**Comment by Hi-Angel** [comment]
> > lldb is fine.
> 
> I'm very surprised to hear this. How does it work?

Good point. I just tested, and it turns out for me, `lldb` has same problem as `gdb`. Here's a session screenshot:

![Screenshot_20200603_013552](https://user-images.githubusercontent.com/7935057/83576491-9a156100-a53a-11ea-8f72-ea96e12dd678.png)

I deleted my prev. comment since it doesn't seem to be relevant now.

@JerrieLau could you please elaborate how you managed to make it work? Perhaps are you using Rust nightly and they added debug info recently? Anything else?

I tested it with lldb version `10.0.0` and rustc version `1.44.0-nightly (20fc02f83 2020-04-20)`

**Comment by JerrieLau** [comment]
@Hi-Angel I just use it in my IDE(CLion).  Here's a screenshot:
![debug](https://user-images.githubusercontent.com/3468079/83655255-c8557800-a5f0-11ea-9e6c-ac3abd536eff.png)

**Comment by Hi-Angel** [comment]
@JerrieLau what you're doing is different. You're basically treating `x` as a struct and unwrapping its members down to a pointer to elements. Both gdb and lldb can do that:

![Screenshot_20200603_183911](https://user-images.githubusercontent.com/7935057/83657464-8cf38300-a5c9-11ea-8a21-8c106addb032.png)

But as a programmer, I don't want to mess around with private internals of Rust types, which I'd need to know, and which can easily be renamed/changed from a Rust version to version.

**Comment by dario23** [comment]
the [intellij-rust plugin](https://plugins.jetbrains.com/plugin/8182-rust) does have some gdb-integration python code though, might be interesting to look at anyways ([licensed under MIT](https://github.com/intellij-rust/intellij-rust/blob/master/LICENSE)).

**Comment by JerrieLau** [comment]
> @JerrieLau what you're doing is different. You're basically treating `x` as a struct and unwrapping its members down to a pointer to elements. Both gdb and lldb can do that:
> 
> ![Screenshot_20200603_183911](https://user-images.githubusercontent.com/7935057/83657464-8cf38300-a5c9-11ea-8a21-8c106addb032.png)
> 
> But as a programmer, I don't want to mess around with private internals of Rust types, which I'd need to know, and which can easily be renamed/changed from a Rust version to version.

Yes. you are right

**Comment by ortem** [comment]
@JerrieLau @Hi-Angel LLDB supports accessing synthetic children via `frame variable <name>` command:
```
(lldb) frame variable vec[0]
(i32) vec[0] = 1
```

See https://lldb.llvm.org/use/variable.html#synthetic-children for more information:

> Unlike many other visualization features, however, the access to synthetic children only works when using frame variable, and is not supported in expression

Unfortunately, just adding `vec[0]` watch in CLion with LLDB doesn't work yet.



@dario23 As for IJ Rust formatters, they're [already upstreamed](https://github.com/rust-lang/rust/pull/72357). But "cannot subscript non-array type" problem with GDB still exists.

**Comment by wesleywiser** [comment]
Visited during wg-debugging triage. I can confirm this is still an issue. Marking `P-medium` as making this work would be pretty useful when debugging Rust programs in gdb.
